### ISSUE TITLE
ENH: mapclassify plots legend as colorbar

### ISSUE THREAD

**Issue by TLouf** [geopandas/geopandas#2154]
For context, I started this reflection in #2012, in which I try to remove the manual legend building for categorical plots. To do so I make a separate, properly labelled matplotlib `Collection` for every category, which allows to build the legend automatically with a simple `ax.legend()` call. This enters into direct conflict with `mapclassify` plots, which may have some data intervals for which there is no geometry to plot. This led me to the following idea:

> Actually I've been thinking, it would make more sense to me if the legend of a plot with a classifying `scheme` was a discrete colorbar, rather than a legend as it is now, which I find a bit ugly and hard to read. This also makes more sense from the point of view of matplotlib, as labelling objects which are not plotted in a legend is not natural, but it is more so in a colorbar. This should probably be discussed in a separate issue, and would entail a big change of behaviour facing the user, so I don't know if it should be changed, but I wanted to know if that made sense to anyone other than me.
> 
> _Originally posted by @TLouf in https://github.com/geopandas/geopandas/issues/2012#issuecomment-932069693_

to which @martinfleis  replied:

> >  it would make more sense to me if the legend of a plot with a classifying scheme was a discrete colorbar
> 
> Generally, yes. But we should allow both options because in some cases, the colorbar is not fully legible. With `explore` we currently support both options so you can pick the one that works better for your use case. I think that in the ideal case, we would support both in static plotting as well. But yeah, that is for another issue (feel free to move these to a new one).
> 
> ![Screenshot 2021-10-01 at 10 37 05](https://user-images.githubusercontent.com/36797143/135599032-0e918eab-95e2-4431-b196-144bce3e9c4b.png)
> 
> _Originally posted by @martinfleis in https://github.com/geopandas/geopandas/issues/2012#issuecomment-932077909_

But `branca`'s colorbar customization is extremely limited compared to `matplotlib`'s. In the given example, one could simply rotate the colorbar to a vertical one with `orientation='vertical'`, format the tick's values correctly with `format`, or render the intervals at the same size with `spacing='uniform'`. The latter is actually the default in matplotlib, but should not be for these plots IMO, as a 'proportional' spacing is often much more telling for whoever reads the plot. When using a legend and not a colorbar, you're kind of always having a uniform spacing, as there's no visual indication of the intervals' sizes.

So my position is that we should always make a colorbar for `mapclassify` plots. Another option is to add a boolean keyword argument  `colorbar` as in `explore()` to let the user choose. In which case, to make #2012 work I'd need to dig back up the code which manually creates the legend for when `colorbar` is set to `False`. 

What do you think?

**Comment by martinfleis** [comment]
I agree that we should default to colorbar. Regarding the categorical legend option, that is for a discussion. Truth it, that one can always get bins via `mapclassify` and get categorical plot manually if that is really needed. It can be a bit cumbersome but if properly documented, it may be enough.
